### PR TITLE
feat: Include end index in proof

### PR DESCRIPTION
**Pull request by zvolin**

### CODE REVIEW

File: src/simple_merkle/proof.rs [hunk 1]
```diff
@@ -30,19 +31,28 @@ where
         root: &M::Output,
         leaf_hashes: &[M::Output],
     ) -> Result<(), RangeProofError> {
-        let tree = MerkleTree::<MemDb<M::Output>, M>::new();
+        let tree = MerkleTree::<NoopDb, M>::new();
         let mut siblings = self.siblings.iter().collect();
 
-        tree.check_range_proof(root, leaf_hashes, &mut siblings, self.start_idx as usize)
+        if leaf_hashes.len() != (self.end - self.start) as usize {
```
Comment (oblique) [inline src/simple_merkle/proof.rs]: `start` and `end` are public, so we don't know if `end` is indeed bigger than `start`. add a check here.

maybe an clean solution is to implement `fn len(&self) -> Result<u32>`. it doesn't need to be public api.

File: src/nmt_proof.rs [hunk 2]
```diff
@@ -52,26 +58,25 @@ impl<M: NamespaceMerkleHasher<Output = NamespacedHash<NS_ID_SIZE>>, const NS_ID_
...
-                start_idx,
-            },
-            ..
-        } = self else {
-            return Err(RangeProofError::MalformedProof)
+        if self.is_of_absence() {
+            return Err(RangeProofError::MalformedProof);
         };
 
+        let proof_len = self.end_idx() - self.start_idx();
```
Comment (oblique) [inline src/nmt_proof.rs]: same as the other review comment

File: src/nmt_proof.rs [hunk 3]
```diff
@@ -39,6 +40,11 @@ impl<M: NamespaceMerkleHasher<Output = NamespacedHash<NS_ID_SIZE>>, const NS_ID_
         raw_leaves: &[impl AsRef<[u8]>],
         namespace: NamespaceId<NS_ID_SIZE>,
     ) -> Result<(), RangeProofError> {
+        let proof_len = self.end_idx() - self.start_idx();
```
Comment (oblique) [inline src/nmt_proof.rs]: same as the other review comment

File: src/simple_merkle/proof.rs (hunk 1 above)
Comment (zvolin) [reply src/simple_merkle/proof.rs]: done

File: src/simple_merkle/proof.rs [hunk 4]
```diff
@@ -53,6 +53,14 @@ where
         self.end
     }
 
+    pub fn range_len(&self) -> usize {
+        if self.start < self.end {
+            (self.end - self.start) as usize
+        } else {
+            0
+        }
```
Comment (oblique) [inline src/simple_merkle/proof.rs:61]: Or you can do this:

```suggestion
        self.end.saturating_sub(self.start)
```

Comment (oblique) [inline src/simple_merkle/proof.rs]: Or you can do this:

```suggestion
        self.end.saturating_sub(self.start)
```

File: src/simple_merkle/proof.rs [hunk 5]
```diff
@@ -30,19 +31,32 @@ where
...
-        self.start_idx
+        self.start
+    }
+
+    pub fn end_idx(&self) -> u32 {
+        self.end
     }
+
+    pub fn range_len(&self) -> usize {
+        self.end.saturating_sub(self.start)
```
Comment (zvolin) [inline src/simple_merkle/proof.rs]:
```suggestion
        self.end.saturating_sub(self.start) as usize
```

**Review by oblique — APPROVED** [review]
LGTM

**Pull request merged**